### PR TITLE
Improve Windows Visual Studio provisioners

### PR DESCRIPTION
**Pull request by vmapetr**
# Description
This PR improves stability for following Windows provisioners:
1. Install-DotnetSDK
1. Install-VS2017
1. Install-VS2019

Most of default `Invoke-WebRequest` occurrences was replace with custom `Start-DownloadWithRetry` function that support retries logic. Also several improvements in syntax was made.

**How it was tested:**
Windows 2016 - https://github.visualstudio.com/virtual-environments/_build/results?buildId=65825&view=results
Windows 2019 - https://github.visualstudio.com/virtual-environments/_build/results?buildId=65826&view=results

#### Related issue:
Issue: https://github.com/actions/virtual-environments-internal/issues/384

## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/win/scripts/Installers/Windows2019/Install-VS2019.ps1 [hunk 1]
```diff
@@ -151,21 +97,19 @@ $WorkLoads = '--allWorkloads --includeRecommended ' + `
...
-$VSBootstrapperURL = 'https://aka.ms/vs/16/release/vs_Enterprise.exe'
-
-$ErrorActionPreference = 'Stop'
+$ReleaseInPath = "Enterprise"
+$BootstrapperName = "vs_${ReleaseInPath}.exe"
+$BootstrapperUrl = "https://aka.ms/vs/16/release/${BootstrapperName}"
 
 # Install VS
-$exitCode = InstallVS -WorkLoads $WorkLoads -Sku $Sku -VSBootstrapperURL $VSBootstrapperURL
+Install-VS -WorkLoads $WorkLoads -BootstrapperName $BootstrapperName -BootstrapperUrl $BootstrapperUrl
```
Comment (miketimofeev) [inline images/win/scripts/Installers/Windows2019/Install-VS2019.ps1:105]: Arguments order is different in VS17 and VS19

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 2]
```diff
@@ -62,6 +63,71 @@ function Install-Binary
...
+
+    $exitCode = $process.ExitCode
+    if ($exitCode -eq 0 -or $exitCode -eq 3010)
+    {
+      Write-Host "Installation successful"
+      return $exitCode
+    }
+    else
+    {
+      Write-Host"Non zero exit code returned by the installation process : $exitCode"
```
Comment (AlenaSviridenko) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:120]: missed space between Write-Host and "

File: images/win/scripts/Installers/Install-DotnetSDK.ps1 [hunk 3]
```diff
@@ -49,22 +52,24 @@ function InstallSDKVersion (
 
 function InstallAllValidSdks()
 {
-    Invoke-WebRequest -Uri 'https://raw.githubusercontent.com/dotnet/core/master/release-notes/releases-index.json' -UseBasicParsing -OutFile 'releases-index.json'
-    $dotnetChannels = Get-Content -Path 'releases-index.json' | ConvertFrom-Json
+    $releasesIndexPath = Start-DownloadWithRetry -Url 'https://raw.githubusercontent.com/dotnet/core/master/release-notes/releases-index.json' -Name 'releases-index.json'
+    $dotnetChannels = Get-Content -Path $releasesIndexPath | ConvertFrom-Json
```
Comment (AlenaSviridenko) [inline images/win/scripts/Installers/Install-DotnetSDK.ps1:56]: I am a bit confused by `$releasesIndexPath` variable. Looks like you are replacing `releases-index.json` with it, but at the same time on the line above you are passing it in raw string. 
Could it be refactored to something like this?
```
$releasesIndexPath = 'releases-index.json'
$releaseUrl = 'https://raw.githubusercontent.com/dotnet/core/master/release-notes/releases-index.json'
Start-DownloadWithRetry -Url $releaseUrl -Name $releasesIndexPath
$dotnetChannels = Get-Content -Path $releasesIndexPath | ConvertFrom-Json
```

File: images/win/scripts/Installers/Windows2019/Install-VS2019.ps1 (hunk 1 above)
Comment (vmapetr) [reply images/win/scripts/Installers/Windows2019/Install-VS2019.ps1]: Done. Thanks!

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 2 above)
Comment (vmapetr) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Fixed. Thanks!

File: images/win/scripts/Installers/Install-DotnetSDK.ps1 (hunk 3 above)
Comment (vmapetr) [reply images/win/scripts/Installers/Install-DotnetSDK.ps1:58]: Fixed. Thanks!

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 4]
```diff
@@ -62,6 +63,71 @@ function Install-Binary
     }
 }
 
+Function Install-VS
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:66]: I think that name `Install-VisualStudio` will be clearer.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 5]
```diff
@@ -62,6 +63,71 @@ function Install-Binary
...
+    .PARAMETER BootstrapperName
+        The Name with which bootstrapper will be downloaded. Required parameter.
+
+    .PARAMETER WorkLoads
+        The string that contain workloads that will be passed to the installer.
+    #>
+  Param
+  (
+    [Parameter(Mandatory)]
+    [String] $BootstrapperName,
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:87]: Can we extract `$BootstrapperName` from `$BootstrapperUrl` like
```
$BootstrapperName = [IO.Path]::GetFileName($BootstrapperUrl)
```

Comment (vmapetr) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Good point, thanks!

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 4 above)
Comment (vmapetr) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Fixed

**Pull request merged**